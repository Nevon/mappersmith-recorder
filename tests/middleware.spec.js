const { install, uninstall, mockClient } = require("mappersmith/test");
const MockDate = require("mockdate");
const { RecorderMiddleware, Session } = require("../");
const MockPersistence = require("./helpers/mockPersistence");
const createClient = require("./helpers/client");
const statusFixture = require("./helpers/fixtures/currentStatus");
const createSubscriberFixture = require("./helpers/fixtures/createSubscriber");

describe("Recorder Middleware", () => {
  beforeEach(() => {
    MockDate.set("2019-06-10T08:00:00.000Z");
    install();
  });

  afterEach(() => {
    uninstall();
    MockDate.reset();
  });

  it("records each request and response", async () => {
    const persistence = new MockPersistence();
    const session = new Session({ persistence, name: "github" });

    const client = createClient({
      middleware: [RecorderMiddleware({ session })]
    });

    mockClient(client)
      .resource("Status")
      .method("current")
      .with({ foo: "bar", baz: 1 })
      .status(200)
      .response(statusFixture);

    mockClient(client)
      .resource("Incidents")
      .method("all")
      .status(500);

    const response = await client.Status.current({ foo: "bar", baz: 1 });

    MockDate.set("2019-06-10T08:00:01.000Z");
    await client.Incidents.all().catch(() => {});
    expect(response.data()).toEqual(statusFixture);

    await session.save();
    session.reset();

    expect(persistence.getFiles()).toMatchSnapshot();
  });

  it("includes post data", async () => {
    const persistence = new MockPersistence();
    const session = new Session({ persistence, name: "github" });

    const client = createClient({
      middleware: [RecorderMiddleware({ session })]
    });

    const requestParams = {
      body: { email: "tyler@statuspage.io", incident: "n4zblpn6p2kq" }
    };
    mockClient(client)
      .resource("Subscribers")
      .method("create")
      .with(requestParams)
      .status(201)
      .response(createSubscriberFixture);

    await client.Subscribers.create(requestParams);

    await session.save();
    expect(
      persistence.getFiles()["github-1560153600000.har"].log.entries[0]._request
        .postData
    ).toMatchSnapshot();
  });
});
