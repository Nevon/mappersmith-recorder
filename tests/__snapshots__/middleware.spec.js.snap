// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recorder Middleware includes post data 1`] = `
PostData {
  "comment": undefined,
  "mimeType": "application/octet-stream",
  "params": Array [],
  "text": "email=tyler%40statuspage.io&incident=n4zblpn6p2kq",
}
`;

exports[`Recorder Middleware records each request and response 1`] = `
Object {
  "github-1560153601000.har": Object {
    "log": Log {
      "browser": undefined,
      "comment": undefined,
      "creator": Creator {
        "comment": undefined,
        "name": "mappersmith-recorder",
        "version": "0.0.1",
      },
      "entries": Array [
        Entry {
          "_request": Request {
            "bodySize": 0,
            "comment": undefined,
            "cookies": Array [],
            "headers": Array [],
            "headersSize": 86,
            "httpVersion": "",
            "method": "GET",
            "postData": PostData {
              "comment": undefined,
              "mimeType": "application/octet-stream",
              "params": Array [],
              "text": "",
            },
            "queryString": Array [
              Pair {
                "comment": undefined,
                "name": "foo",
                "value": "bar",
              },
              Pair {
                "comment": undefined,
                "name": "baz",
                "value": "1",
              },
            ],
            "url": "https://kctbh9vrtdwd.statuspage.io/api/v2/status.json?foo=bar&baz=1",
          },
          "_response": Response {
            "bodySize": 215,
            "comment": undefined,
            "content": Content {
              "comment": undefined,
              "compression": undefined,
              "encoding": undefined,
              "mimeType": "application/json",
              "size": 215,
              "text": "{\\"page\\":{\\"id\\":\\"kctbh9vrtdwd\\",\\"name\\":\\"GitHub\\",\\"url\\":\\"https://www.githubstatus.com\\",\\"time_zone\\":\\"Etc/UTC\\",\\"updated_at\\":\\"2019-06-09T08:06:53.070Z\\"},\\"status\\":{\\"indicator\\":\\"none\\",\\"description\\":\\"All Systems Operational\\"}}",
            },
            "cookies": Array [],
            "headers": Array [
              Pair {
                "comment": undefined,
                "name": "content-type",
                "value": "application/json",
              },
            ],
            "headersSize": 34,
            "httpVersion": "HTTP/1.1",
            "redirectURL": "",
            "status": 200,
            "statusText": "OK",
          },
          "cache": Object {
            "afterRequest": null,
            "beforeRequest": null,
            "comment": undefined,
          },
          "comment": undefined,
          "connection": undefined,
          "pageref": undefined,
          "request": Request {
            "bodySize": 0,
            "comment": undefined,
            "cookies": Array [],
            "headers": Array [],
            "headersSize": 86,
            "httpVersion": "",
            "method": "GET",
            "postData": PostData {
              "comment": undefined,
              "mimeType": "application/octet-stream",
              "params": Array [],
              "text": "",
            },
            "queryString": Array [
              Pair {
                "comment": undefined,
                "name": "foo",
                "value": "bar",
              },
              Pair {
                "comment": undefined,
                "name": "baz",
                "value": "1",
              },
            ],
            "url": "https://kctbh9vrtdwd.statuspage.io/api/v2/status.json?foo=bar&baz=1",
          },
          "response": Response {
            "bodySize": 215,
            "comment": undefined,
            "content": Content {
              "comment": undefined,
              "compression": undefined,
              "encoding": undefined,
              "mimeType": "application/json",
              "size": 215,
              "text": "{\\"page\\":{\\"id\\":\\"kctbh9vrtdwd\\",\\"name\\":\\"GitHub\\",\\"url\\":\\"https://www.githubstatus.com\\",\\"time_zone\\":\\"Etc/UTC\\",\\"updated_at\\":\\"2019-06-09T08:06:53.070Z\\"},\\"status\\":{\\"indicator\\":\\"none\\",\\"description\\":\\"All Systems Operational\\"}}",
            },
            "cookies": Array [],
            "headers": Array [
              Pair {
                "comment": undefined,
                "name": "content-type",
                "value": "application/json",
              },
            ],
            "headersSize": 34,
            "httpVersion": "HTTP/1.1",
            "redirectURL": "",
            "status": 200,
            "statusText": "OK",
          },
          "serverIPAddress": undefined,
          "startedDateTime": "2019-06-10T08:00:00.000Z",
          "time": -1,
          "timings": Object {
            "blocked": -1,
            "comment": undefined,
            "connect": -1,
            "dns": -1,
            "receive": -1,
            "send": -1,
            "wait": -1,
          },
        },
        Entry {
          "_request": Request {
            "bodySize": 0,
            "comment": undefined,
            "cookies": Array [],
            "headers": Array [],
            "headersSize": 75,
            "httpVersion": "",
            "method": "GET",
            "postData": PostData {
              "comment": undefined,
              "mimeType": "application/octet-stream",
              "params": Array [],
              "text": "",
            },
            "queryString": Array [],
            "url": "https://kctbh9vrtdwd.statuspage.io/api/v2/incidents.json",
          },
          "_response": Response {
            "bodySize": 0,
            "comment": undefined,
            "content": Content {
              "comment": undefined,
              "compression": undefined,
              "encoding": undefined,
              "mimeType": "application/octet-stream",
              "size": 0,
              "text": "",
            },
            "cookies": Array [],
            "headers": Array [],
            "headersSize": 4,
            "httpVersion": "HTTP/1.1",
            "redirectURL": "",
            "status": 500,
            "statusText": "Internal Server Error",
          },
          "cache": Object {
            "afterRequest": null,
            "beforeRequest": null,
            "comment": undefined,
          },
          "comment": undefined,
          "connection": undefined,
          "pageref": undefined,
          "request": Request {
            "bodySize": 0,
            "comment": undefined,
            "cookies": Array [],
            "headers": Array [],
            "headersSize": 75,
            "httpVersion": "",
            "method": "GET",
            "postData": PostData {
              "comment": undefined,
              "mimeType": "application/octet-stream",
              "params": Array [],
              "text": "",
            },
            "queryString": Array [],
            "url": "https://kctbh9vrtdwd.statuspage.io/api/v2/incidents.json",
          },
          "response": Response {
            "bodySize": 0,
            "comment": undefined,
            "content": Content {
              "comment": undefined,
              "compression": undefined,
              "encoding": undefined,
              "mimeType": "application/octet-stream",
              "size": 0,
              "text": "",
            },
            "cookies": Array [],
            "headers": Array [],
            "headersSize": 4,
            "httpVersion": "HTTP/1.1",
            "redirectURL": "",
            "status": 500,
            "statusText": "Internal Server Error",
          },
          "serverIPAddress": undefined,
          "startedDateTime": "2019-06-10T08:00:01.000Z",
          "time": -1,
          "timings": Object {
            "blocked": -1,
            "comment": undefined,
            "connect": -1,
            "dns": -1,
            "receive": -1,
            "send": -1,
            "wait": -1,
          },
        },
      ],
      "pages": Array [],
      "version": "1.2",
    },
  },
}
`;
